lesson22/src/js/main.js

```diff
@@ -0,0 +1,213 @@
+        str:error
+    });
+    element.appendChild(errorMessage);
+}
+
+const fetchData = async (endpointURL,element) => {
+    const response = await fetch(endpointURL);
+    if(!response.ok){
+        console.error(`${response.status}:${response.statusText}`);
+        renderErrorMessage("Communication with the server is broken.",element);
+        return;
+    }
+    return await response.json();
+}
+
+const fetchContentsData = (endPointURL,element,ms) => new Promise(resolve => setTimeout(() => resolve(fetchData(endPointURL,element)),ms));
+
+const table = document.getElementById("js-table");
+
+const initUsersData = async () => {
+    loading.showLoading(table);
+    try{
+        const json = await fetchContentsData(endPointURL.usersData,table,500);
+        const usersData = json.data;
+
+        if (!usersData.length) {
+            renderErrorMessage("No user.",table);
+            return;
+        }
+        renderTableElements(usersData);
+        sortTableBody(usersData);
+    }catch(error){
+        console.error(error);
+    }finally{
+       loading.removeLoading(table);
+    }
+}
+
+const renderTableElements = usersData => {
+    renderTableHeader();
+    renderTableBody(usersData);
+}
+
+const usersTableColumn = {
+    "id"     : "ID",
+    "name"   : "名前",
+    "gender" : "性別",
+    "age"    : "年齢"
+}
+
+const renderTableHeader = () => {
+    const thead = document.createElement("thead");
+    const tr = document.createElement("tr");
+    thead.className = "bg-slate-500";
+
+    for (const [userKey, userValue] of Object.entries(usersTableColumn)) {
+        const th = createAttributedElements({
+            tag:"th",
+            valuesByAttributes:{
+                class:"text-sm text-white px-6 py-4",
+            },
+            str:userValue
+        });
+
+        tr.appendChild(th);
+
+        if (sortCategories.includes(userKey)) {
+            th.appendChild(createSortButtons(userKey));
+        }
+    };
+
+    table.appendChild(thead).appendChild(tr);
+}
+
+const renderTableBody = usersData => {
+    const tbody = document.createElement("tbody");
+    tbody.id = "js-usersTableBody";
+
+    for(const user of usersData){
+        const fragment = document.createDocumentFragment();
+        const tr = document.createElement("tr");
+
+        Object.keys(usersTableColumn).forEach((column) => {
+            const td = createAttributedElements({
+                tag:"td",
+                valuesByAttributes:{
+                    class:"border border-gray-300 px-4 py-2"
+                },
+                str:user[column]
+            })
+            fragment.appendChild(tr).appendChild(td);
+        });
+
+        table.appendChild(tbody).appendChild(fragment);
+    }
+}
+
+const sortButtonAttributes = [
+    {state:"asc",src:"../img/asc.svg", alt:"asc-image"},
+    {state:"desc",src:"../img/desc.svg",alt:"desc-image"},
+    {state:"both",src:"../img/both.svg",alt:"both-image"}
+]
+
+const sortCategories = ["id", "age"];
```

Comment:
IMO
I think it would be easier later to manage the object together with the column constants.
Because you will have to add buttons one by one as they are added in the future.

```js
const sortCategories = ["id", "age", "hogehoge","hugahuga"......];
```

For example...

```js
const tableColumnConfig = {
  id: {
    value: "ID",
    hasSort: true
  },
  name: {
    value: "名前",
    hasSort: false
  },
  gender: {
    value: "性別",
    hasSort: true
  },
  age: {
    value: "年齢",
    hasSort: false
  }
};
```
P.S
If `hasSort` is ture, you will need to rewrite the code to display the sort button.
It's going to be a big revision and it's a crazy detail, so just think of it as this kind of idea.

Reply:
Thanks for your comment.

You're right, it would be easier to manage the sorting part in one object. 
I'll see if I can figure out a way to make it an object!

Reply:
I've decided to use an object to manage the sorting elements.
 I'm a little unsure if this is the right way to do it.

https://github.com/nakagawamai/js-lesson/pull/24/commits/a50b9f3391b69a37cadaf81be7931295d0a56d23